File: 41-first-missing-positive/41-first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        # 정규화 (가능하지 않은 숫자 다 -1로)
        for i in range(len(nums)):
            if nums[i] <= 0 or nums[i] > len(nums):
                nums[i] = -1

        # 각 원소를 라벨로 하여 건너가며 제거, 즉 0으로 변환 (꼬리물기)
        for i in range(len(nums)):
            next: int = nums[i]
            # 원소가 다음을 가리키면 다음 위치 저장, 아니면 -1
            if next > 0:
                store: int = next
            else:
                store: int = -1

            while store != -1:
                curr: int = store
                next = nums[curr-1]
                # 원소가 다음을 가리키면 다음 위치 저장
                if next > 0:
                    store = next
                else:
                    store = -1
                nums[curr-1] = 0

        # 훑으며 방문하지 않은 첫번째 구간 위치 반환
        for i in range(len(nums)):
            if nums[i] != 0:
                return i+1
        return len(nums)+1

# testcases
s: Solution = Solution()
assert s.firstMissingPositive([1, 2, 0]) == 3
assert s.firstMissingPositive([3, 4, -1, 1]) == 2
assert s.firstMissingPositive([7, 8, 9, 11, 12]) == 1
assert s.firstMissingPositive([1, 1]) == 2

File: 41-first-missing-positive/41-first-missing-positive.rs
impl Solution {
    pub fn first_missing_positive(nums: Vec<i32>) -> i32 {
        let mut nums = nums.clone();
        let n = nums.len();
        // make numbers which is out of range to -1
        for x in nums.iter_mut() {
            *x = {
                if *x <= 0 || *x > (n as i32) {
                    -1
                }
                else { *x }
            };
        }

        // use each elements as a postional index, make elements as 0
        for i in 0..n {
            let mut next = *nums.get(i).unwrap();
            // if element is point to next then store next position, or store -1
            let mut store = if next > 0 { next } else { -1 };

            while store != -1 {
                let curr = store;
                next = *nums.get((curr - 1) as usize).unwrap();
                // if element points next then store next position
                store = if next > 0 { next } else { -1 };
                *nums.get_mut((curr-1) as usize).unwrap() = 0;
            }
        }

        // return first postition which didn't checked (nonzero)
        for (i, x) in nums.iter().enumerate() {
            if *x != 0 { return (i+1) as i32; }
        }
        (n + 1) as i32
    }
}